improve code qualtiy of unit_custom_weapons_behaviors

efrec opened the pull request:
### Work done

This one gadget file is chaos and seeing it again was all it took to do this.

- Refactor existing unit_custom_weapons_behaviours for better practices
- Refactor existing weapon customparams to use better param names
- Add error checks, simplifies setup of weapon behaviors
- Some CPU, memory, and allocation improvements
- Remove stale todos that are obvious enough in code

Please wait to merge but feel free to review for code practices. It will take me a bit to test that each of these works exactly the same as before (cruise, retarget, sector_fire, cannon/torpedo water pen) via ingame validation. Won't be long.

One notable change that maybe testing will explain to me (so will revert): I made torpedo's velocity reduction of water impact consistent across target depths. The previous behavior, though, genuinely did not make sense to me.

efrec commented:
Made, so fixed, a couple of mistakes. Everything is tested, bug free, and a bit lighter weight than before. Plus I can look at it without being bothered.

ZephyrSkies7 commented:
@WatchTheFort @sprunk please review

WatchTheFort commented:
I'm too busy for the next week to give this a proper review, but from a glance I would love it if the cryptic three letter variables could be given full names, e.g. `pvx` -> `projectileVelocityX`.

I am super not a fan of the `speceffect_when = "ypos<=0",` custom params, because they are encoding programmatic logic inside a string.

efrec commented:
Noted, will fix. Thanks.

sprunk commented:
Would be easier to review if fixes were squashed into their primary commit and cosmetic commits (no logic changes) were frontloaded.

efrec commented:
It did start that way, yeah. I'll fix up these commits.

efrec commented:
Draft-ed this since I made it basically un-reviewable until I remake the commits.

sprunk commented:
Could also put the first handful of commits in a separate tiny PR so we could nibble away at it in chunks, since nobody seems to too eager to review the entire rewrite.

efrec commented:
Wish I'd checked here xd I've been staring at it for three weeks, sorry

saurtron commented on luarules/gadgets/unit_custom_weapons_behaviours.lua:
```diff
@@ -2,334 +2,375 @@ local gadget = gadget ---@type Gadget
+end
 
+weaponSpecialEffect.split = function(params, projectileID)
+	if isProjectileFalling(projectileID) then
```
what's up with these isProjectileInWater and isProjectileFalling methods, I don't see them defined anywhere.

efrec replied:
missed them in the redo, should be fixed

efrec replied:
I'm loading into a game to test now, just did this push over a lunch break so didn't test.

saurtron replied:
K, test a bit pls :D

efrec replied:
Done and dusted.

saurtron reviewed (changes requested):
Tried to test the gadget by creating `armlance`, `armsptk` and `armseap` units, but I get errors with the method isProjectileInWater.

Don't see those defined (isProjectileInWater or isProjectileFalling) but they also were in use before and didn't exist so not sure what's going on.

`Error: [LuaRules::RunCallInTraceback] error=2 (LUA_ERRRUN) callin=GameFrame trace=[Internal Lua error: Call failure] [string "LuaRules/Gadgets/unit_custom_weapons_behaviours.lua"]:309: attempt to call global 'isProjectileInWater' (a nil value)`

saurtron commented on luarules/gadgets/unit_custom_weapons_behaviours.lua:
```diff
@@ -2,334 +2,414 @@ local gadget = gadget ---@type Gadget
+	end
+end
+
+function gadget:ProjectileCreated(projectileID, proOwnerID, weaponDefID)
```
ProjectileCreated and ProjectileDestroyed are used but don't see SetWatchProjectile being used.

saurtron commented on luarules/gadgets/unit_custom_weapons_behaviours.lua:
```diff
@@ -2,334 +2,418 @@ local gadget = gadget ---@type Gadget
 
+-- Water penetration with retargeting (torpedo)
 
+do
```
why this `do / end` block?

efrec replied:
I do not have a strong reason. I did not want to put the two function references in the same scope of the rest of the script.

saurtron replied:
Ok, I see it does make the two local speedups private so I guess it's ok. It's not standard technique and also not sure what the implications could be about capturing outer scope, maybe @sprunk knows if it's a good practice or not.

My experience with closures inside functions is they create a lot of overhead, but maybe not at the top scope like this.

saurtron commented on luarules/gadgets/unit_custom_weapons_behaviours.lua:
```diff
@@ -2,334 +2,414 @@ local gadget = gadget ---@type Gadget
+		do
+			local targetType, target = spGetProjectileTarget(projectileID)
+			if targetType == targetedUnit then
+				local _; -- sink for unused values
```
why a semicolon here instead of comma? also this is inside a `do / end` block with no real use afaics.

efrec replied:
The current code assigns to a global named `_`, so `local _;` declares it as a local, instead. Underscores for unused variables are common practice but there's no actual, special behavior behind them.

I might use `do/end` excessively just to organize blocks.

How about:

```lua
local targetX, targetY, targetZ
local targetType, target = spGetProjectileTarget(projectileID)

if targetType == targetedUnit then
	local _; -- declare a local sink var for unused values
	_, _, _, targetX, targetY, targetZ = spGetUnitPosition(target, false, true)
elseif targetType == targetedGround then
	targetX, targetY, targetZ = target[1], target[2], target[3]
end
```

saurtron replied:
yeah like this is more understandable.

ok I see you make `local _;` just so you can mix with the previously declared locals.

I do agree `_` is ok and standard procedure for unused variables, what's weird here is the `local _` and specially the semicolon since I think I never seen a semicolon before in lua... thought that made it continue on the next line XD.

efrec replied:
This is, indeed, BAR's only line terminator used in Lua

saurtron commented on luarules/gadgets/unit_custom_weapons_behaviours.lua:
```diff
@@ -2,334 +2,418 @@ local gadget = gadget ---@type Gadget
+	getSpawnParams = function(params, projectileID)
+		local spawnDefID = params.speceffect_def
+
+		local cache = spawnCache
```
Some remarks about the whole dealing with spawnCache.

First of all, it's not really a cache is it, it's more of a single table "mempool", not a big deal to call it cache here although a bit misleading since initial thought is the values will be reused later on (they're just overwritten to forward into spSpawnProjectile).

Then, you assign it to `local cache` but then still using both cache and `spawnCache` in the next lines.

I think doing `local spawnArgs = spawnCache` (or spawnParams or similar) and using it just as that inside the function, then when it gets returned from getSpawnParams at the calling location it should probably be called `spawnArgs` (or similar) too instead of cache.

I thought about `params` but that's taken in the function, and spawnParams could make it confusing since the method is called getSpawnParams but it returns more stuff, that's why i suggest spawnArgs but I'd be happy if you find smth better too :D.

efrec replied:
It's barely effective as a cache. Mostly it pulls out some duplicate code, and a shared table was better than, idk, a mini-factory.

I found a matching `---@class ProjectileParams` in the new luadocs, so I can add some annotations, and will change from "params" to "args".

saurtron commented on luarules/gadgets/unit_custom_weapons_behaviours.lua:
```diff
@@ -2,334 +2,418 @@ local gadget = gadget ---@type Gadget
-	applyingFunctions.torpwaterpenretarget = function (proID)
-		return false
+	Spring.DeleteProjectile(projectileID)
+	Spring.SpawnCEG(params.splitexplosionceg, cache.pos[1], cache.pos[2], cache.pos[3])
```
might be a bit more efficient to do like `Spring.SpawnCEG(params.splitexplosionceg, unpack(cache.pos))`

(if it works but I think it should, didn't try it)

saurtron commented on luarules/gadgets/unit_custom_weapons_behaviours.lua:
```diff
@@ -2,334 +2,418 @@ local gadget = gadget ---@type Gadget
-	--fake function
-	applyingFunctions.torpwaterpenretarget = function (proID)
-		return false
+	Spring.DeleteProjectile(projectileID)
```
I'd add more speedups for all the critical per-projectile effect functions in the module, like Spring.DeleteProjectile, or Spring.SpawnCEG here, also Spring.SpawnProjectile elsewhere and maybe some other.

saurtron reviewed (changes requested):
Will still need a bit more review tomorrow, and also some testing (I did test all non-legion units here but just quickly), but overall I'm liking the code and module status and things seem to work.

It does seem to me like a good improvement.

I specially like how you removed the `when` customParam since all effects were always using the same `when`. Also some of the tricks here, and generally how the whole module seems to be much more simple now. There's a couple var naming issues that make some things difficult to follow initially but I'm sure we (well... you :D) can sort them out, I'll point them out pls bear with me about that.

Not sure if good or bad, but some of the techniques seem very personal, and novel inside the bar codebase (like the `do/end` scope/closures, the metatable trick for effects, and the mixed local/placeholder), but overall I like those too, although in some places I'm not totally sure if they really belong or just overused (specially the `do` blocks seem a bit distracting with little gain other than making few locals private)... It's funny since I also can't see how you could do without them but then again never before felt the need for them or seem them much around xd.

Will approve once small remaining things will be fixed and I'm happy with final review, thanks for your time working on this.

saurtron commented on luarules/gadgets/unit_custom_weapons_behaviours.lua:
```diff
@@ -2,334 +2,415 @@ local gadget = gadget ---@type Gadget
+-- Initialization --------------------------------------------------------------
+
+local weaponCustomParamKeys = {} -- [effect] = { key => conversion function }
+local weaponSpecialEffect = {}
```
Regarding the naming of this variable (`weaponSpecialEffect`), I liked how before it was called `applyingFunctions`,  that made it really straightforward to understand.

Maybe could still have that name or smth in that line (`weaponEffectFunctions` maybe?), otherwise something has to be done since also now having both `weaponEffect` and `weaponSpecialEffect` tables is a bit confusing tbh :D.

In the end one has the pure functions, and the other one the closure of "function + args" through the metatable, but that's not evident until you reach the end of the module, and makes the initial read a bit confusing.

If those names can be picked right I'm sure it'll help, sorry about looking so picky about names but I think it will really help.

efrec replied:
Hopefully I picked correctly here, I removed the reference to "weapon" on the base effect table to strengthen the difference between the general effects and their weaponDef versions with baked-in params.

saurtron commented on luarules/gadgets/unit_custom_weapons_behaviours.lua:
```diff
@@ -2,334 +2,429 @@ local gadget = gadget ---@type Gadget
+			local _; -- declare a local sink var for unused values
+			_, _, _, targetX, targetY, targetZ = spGetUnitPosition(target, false, true)
+		elseif targetType == targetedGround then
+			targetX, targetY, targetZ = target[1], target[2], target[3]
```
I think can unpack(target) here.

efrec replied:
Yeah this is a micro-opt because unpack is slightly slower, and cruise is the slowest of the effects (every frame, 8 calls to engine, etc). But it is weaker code style. Idk, would prefer to keep, it's not confusing as it is.

efrec replied:
K I did the rest. Can I be slick about this, like

```lua
local function unpack3(tbl)
   return tbl[1], tbl[2], tbl[3]
end
```

or is that actually more annoying

saurtron replied:
yeah, no, was just meaning to use unpack as possible perf improvement, but if it's actually slower no reason to use it imo, sorry about that.

saurtron commented on luarules/gadgets/unit_custom_weapons_behaviours.lua:
```diff
@@ -2,334 +2,429 @@ local gadget = gadget ---@type Gadget
+				if ownerTargetType == 1 then
+					spSetProjectileTarget(projectileID, ownerTarget, targetedUnit)
+				elseif ownerTargetType == 2 then
+					spSetProjectileTarget(projectileID, ownerTarget[1], ownerTarget[2], ownerTarget[3])
```
Can probably unpack(ownerTarget) here.

saurtron commented on luarules/gadgets/unit_custom_weapons_behaviours.lua:
```diff
@@ -2,334 +2,429 @@ local gadget = gadget ---@type Gadget
-			end
+end
+
+-- Cruise
```
Maybe its asking much, but would be great to have a small effect description and paramKeys description here for each effect. Might be easy for you now that it's fresh and will help a lot in getting the module documented.

efrec replied:
You're fine 😆 It's my fault for shifting gears from "I hate this file" to "I am going to fix this file".

saurtron commented on luarules/gadgets/unit_custom_weapons_behaviours.lua:
```diff
@@ -2,334 +2,429 @@ local gadget = gadget ---@type Gadget
-		Spring.SpawnCEG(infos.waterpenceg, px, py, pz,0,0,0,0,0)
-		Spring.DeleteProjectile(proID)
+			if effectName then
+				weaponDefEffect[weaponDefID] = setmetatable(effectParams, metatables[effectName])
```
Would be great to assign the `effectFunction` directly instead of metatable for those effects not having any `weaponCustomParamKeys`. That way the specialEffectFunction can skip the params arguments when not used and will be more efficient.

If so, might also be better to switch arguments for the specialEffectFunctions so params goes second and looks less confusing when not present.

saurtron commented on luarules/gadgets/unit_custom_weapons_behaviours.lua:
```diff
@@ -2,334 +2,429 @@ local gadget = gadget ---@type Gadget
+	spSpawnProjectile(spawnDefID, spawnParams)
+end
+
+specialEffectFunction.cannonwaterpen = function(projectileID)
```
Looks like `params` is missing here and in forwarding to `cannonWaterPen` below.

efrec commented:
I hope this is done enough to say thanks for some good changes here and for sticking out the change in direction.

saurtron reviewed (changes requested):
Tested all weapons and most look fine but the `torpedowaterpenretarget` seems broken, it just dissapears on first hit into the water (tested it by changing `armseap` to `torpwaterpenretarget`).

saurtron commented:
> One notable change that maybe testing will explain to me (so will revert): I made torpedo's velocity reduction of water impact consistent across target depths. The previous behavior, though, genuinely did not make sense to me.


Regarding the velocity change you removed for different depths, seems generally subs will be below -10, while ships will be -1 (kind of evident but I checked to be sure XD).

Previously it wasn't reducing speed for ships and structures, while it was for subs. So, this is a bit of a nerf for ships and structures? I think it should either not reduce speed in x and z at all, or keep the same logic as before.

Personally I'd try to keep the logic same as before to avoid issues (since this is a refactoring PR), but then propose a different PR with removing that special behaviour if you think it's the way to go, so GDT and other reviewers can take a closer look at that. Also we can't test all possible scenarios so it's possible it would introduce regressions.

efrec commented:
Yeah I put that back as `if positionY and positionY < -10 then` after sorting through the details, and added the comment above to justify it a bit. I could put that ships are "assumed to be at a depth of -1" or something, then, since that is pretty specific info to know.

efrec commented:
Noted, fixed.

I removed all remaining minor changes, like ending retarget early, except for torpwaterpenretarget. There's just no way it is intended to repeat its waterpen behavior each frame.

I tested an armseap and did not see the projectile disappear or be deleted. There's no code between retarget or torpwaterpen that can delete a projectile, and no real difference in how TorpedoProjectile speed and target updates work vs other projectiles. Stumped. I tested before and after these changes from today.

saurtron commented:
> I tested an armseap and did not see the projectile disappear or be deleted. There's no code between retarget or torpwaterpen that can delete a projectile, and no real difference in how TorpedoProjectile speed and target updates work vs other projectiles. Stumped. I tested before and after these changes from today.

Ok, yeah I tested with current and previous code, and now seems to work fine, I think customweapon was just not doing anything so the torpedo was going to the water and then into the bottom of the ocean, did look like dissapearing but it was just not going towards the target. It's all good now.

how it was looking for the record:

![torpretarget3](https://github.com/user-attachments/assets/2cdc204d-06bc-4140-a31e-19457c285762)

saurtron commented on luarules/gadgets/unit_custom_weapons_behaviours.lua:
```diff
@@ -2,334 +2,467 @@ local gadget = gadget ---@type Gadget
 
+specialEffectFunction.cannonwaterpen = function(params, projectileID)
+	if isProjectileInWater(projectileID) then
+		cannonWaterPen(projectileID)
```
I think it needs to forward params here.

efrec replied:
Dang I thought I fixed that when you had checked it earlier.

efrec commented:
It looks like the gadget may have unloaded. I picked around for a failure point but came up with only an unrelated one.

saurtron commented:
how do you test watercannonpen btw, not sure what unit to tweak?

saurtron reviewed (approved):
All right, looking good :D

In case you want to keep working on this after this PR, I feel like it could be greatly improved by adding a nextCheckFrame tracking system for the projectiles, similar to the one `unit_projectile_overrange.lua` uses.

It keeps a dict with next frame number to check, with arrays of projIDs, so they go through "the check in N frames" phase, to the "always check phase". This is so, since when created you're often just waiting until it hits the ground/water, and that can be somewhat predicted (see `calculateFlightFrames()` function there, maybe that one should go in a common module so it can be reused).

Also, maybe there are a few things you wanted to improve or change, in case you don't want to work on those I suggest you create an issue with those too, now that it's fresh on your mind.

Anyways this is a refactor PR and as such I think its ready, you've done a great job, thx for bearing with my review :D.

efrec commented:
*phew* thanks much!

There are many small improvements, like how cruise checks a minimum y velocity, then just sets its y velocity to any value. And retarget does not always work. I'll put off a big review for a bit.

efrec commented:
@WatchTheFort Ready, and has been tested.
